app/src/main/java/com/example/weatherapp/models/DetailsActivity.kt

```diff
@@ -21,21 +23,19 @@ class DetailsActivity: AppCompatActivity() {
 
     private val city: String by lazy { intent.getStringExtra("Temperature_Info").toString() }
 
-    private fun initialiseComponentInfo(): Unit{
-        locationName.text = city
-    }
+    private val detailsViewModel: WeatherDetailsViewModel by viewModels()
 
-    private fun setComponentInfo(temperatureInfo: TemperatureResult?){
-        if(temperatureInfo==null){
-            return
-        }
-        locationHumidity.text = temperatureInfo.humidity
-        locationTemperature.text = temperatureInfo.temp.toString() + " °F"
-        locationDescription.text = temperatureInfo.description
-        setTemperaturePhoto(temperatureInfo.icon)
+    private fun setComponentInfo(uiState: WeatherDetailsUiState){
+        locationName.text = uiState.weatherInformation?.address
+        val latestWeatherResultIndex: Int = 0
+        val weatherInformation = uiState.weatherInformation?.days?.get(latestWeatherResultIndex)
```

Comment:
Nice one moving away from the `TemperatureResult` here.

Another improvement is thinking of the View layer as being very dumb. It shouldn't know much, and shouldn't make any decisions.

With that in mind, the View here could take some of this data already mapped and formatted from the ViewModel. This is the kind of stuff the ViewModel is good at. And, when you add tests to your code, we usually do test the ViewModel, and having more of that logic in there, the more tested the code will be.